pkg/controller/sidecarset/sidecarset_utils.go

```diff
@@ -101,3 +112,182 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
 		status.UpdatedPods != sidecarSet.Status.UpdatedPods ||
 		status.ReadyPods != sidecarSet.Status.ReadyPods
 }
+
+// add this cache to avoid be influenced by informer cache latency when controller try to count maxUnavailable
+type updatedPodCache struct {
+	lock sync.RWMutex
+	// key is in the format: sidecarset name/pod namespace/pod name
+	// value is sidecarset hash
+	podSidecarUpdated map[string]string
+}
+
+func (u *updatedPodCache) set(key, hash string) {
+	u.lock.Lock()
+	u.podSidecarUpdated[key] = hash
+	u.lock.Unlock()
+}
+
+func (u *updatedPodCache) reset() {
+	u.lock.Lock()
+	u.podSidecarUpdated = make(map[string]string)
+	u.lock.Unlock()
+}
+
+// return true means: sidecar of pod is updated
+func (u *updatedPodCache) isUpdated(key, curHash string) bool {
+	u.lock.RLock()
+	hash := u.podSidecarUpdated[key]
+	u.lock.RUnlock()
+
+	if hash == curHash {
+		return true
+	}
+	if hash != "" && hash != curHash {
+		u.reset()
+		return false
+	}
+	return false
+}
+
+// available definition:
+// 1. image in pod.spec and pod.status is exactly the same
+// 2. pod is ready
+func getUnavailableNumber(sidecarSet *appsv1alpha1.SidecarSet, pods []*corev1.Pod) (int, error) {
+	var unavailableNum int
+	for _, pod := range pods {
+		// in case of informer cache latency
+		podInCacheUpdated := updateCache.isUpdated(
```

Comment:
Can you explain more why we need this updateCache?

Reply:
sure. even if controller updated pod in the last syncLoop, the pod might still be considered as not updated (happens when informer cache doen't update in time) next time, which will cause maxUnavailable=0